airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/StatePersistence.java

```diff
@@ -0,0 +1,323 @@
+                      final String namespace,
+                      final StateType stateType,
+                      final JsonNode state) {
+    if (state != null) {
+      boolean hasState = ctx.selectFrom(STATE)
+          .where(
+              STATE.CONNECTION_ID.eq(connectionId),
+              isNullOrEquals(STATE.STREAM_NAME, streamName),
+              isNullOrEquals(STATE.NAMESPACE, namespace))
+          .fetch().isNotEmpty();
+
+      final JSONB jsonbState = JSONB.valueOf(Jsons.serialize(state));
+      final OffsetDateTime now = OffsetDateTime.now();
+
+      if (!hasState) {
+        ctx.insertInto(STATE)
+            .columns(
+                STATE.ID,
+                STATE.CREATED_AT,
+                STATE.UPDATED_AT,
+                STATE.CONNECTION_ID,
+                STATE.STREAM_NAME,
+                STATE.NAMESPACE,
+                STATE.STATE_,
+                STATE.TYPE)
+            .values(
+                UUID.randomUUID(),
+                now,
+                now,
+                connectionId,
+                streamName,
+                namespace,
+                jsonbState,
+                convertStateType(stateType))
+            .execute();
+
+      } else {
+        ctx.update(STATE)
+            .set(STATE.UPDATED_AT, now)
+            .set(STATE.STATE_, jsonbState)
+            .where(
+                STATE.CONNECTION_ID.eq(connectionId),
+                isNullOrEquals(STATE.STREAM_NAME, streamName),
+                isNullOrEquals(STATE.NAMESPACE, namespace))
+            .execute();
+      }
+
+    } else {
+      // If the state is null, we remove the state instead of keeping a null row
+      ctx.deleteFrom(STATE)
+          .where(
+              STATE.CONNECTION_ID.eq(connectionId),
+              isNullOrEquals(STATE.STREAM_NAME, streamName),
+              isNullOrEquals(STATE.NAMESPACE, namespace))
+          .execute();
+    }
+  }
+
+  /**
+   * Helper function to handle null or equal case for the optional strings
+   *
+   * We need to have an explicit check for null values because NULL != "str" is NULL, not a boolean.
+   *
+   * @param field the targeted field
+   * @param value the value to check
+   * @return The Condition that performs the desired check
+   */
+  private static Condition isNullOrEquals(final Field<String> field, final String value) {
+    return value != null ? field.eq(value) : field.isNull();
+  }
+
+  /**
+   * Get the StateType for a given list of StateRecords
+   *
+   * @param connectionId The connectionId of the records, used to add more debugging context if an
+   *        error is detected
+   * @param records The list of StateRecords to process, must not be empty
+   * @return the StateType of the records
+   * @throws IOException If StateRecords have inconsistent types
+   */
+  private io.airbyte.db.instance.configs.jooq.generated.enums.StateType getStateType(final UUID connectionId, final List<StateRecord> records) {
```

Comment:
i think a few methods in this class can be `static`. i caught this one but there might be others. as a best practice we generally set methods to be static when possible. it is always fine to remove that keyword if we need to change or mock it.